### PR TITLE
[mlir][llvm] Add icmp folder

### PR DESCRIPTION
**Pull request by gysit**
This revision adds a simple icmp folder that performs the following folds to the LLVM dialect icmp op:
 - cmpi(eq/ne, x, x) -> true/false
 - cmpi(eq/ne, alloca, null) -> false/true
 - cmpi(eq/ne, null, alloca) -> cmpi(eq/ne, alloca, null)

### CODE REVIEW

**Comment by zero9178** [comment]
Looks good!

Since you've gone through the effort of making this work with vector types as well, could you add at least one test that uses vector input types?

**Comment by gysit** [comment]
Good point, the constant materialization did indeed not support vectors!

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp [hunk 1]
```diff
@@ -175,6 +175,42 @@ ParseResult FCmpOp::parse(OpAsmParser &parser, OperationState &result) {
   return parseCmpOp<FCmpPredicate>(parser, result);
 }
 
+/// Returns a scalar or vector boolean attribute of the given type.
+static Attribute getBoolAttribute(Type type, MLIRContext *ctx, bool value) {
+  auto boolAttr = BoolAttr::get(ctx, value);
+  ShapedType shapedType = llvm::dyn_cast_or_null<ShapedType>(type);
```
Comment (zero9178) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:181]:
```suggestion
  ShapedType shapedType = dyn_cast<ShapedType>(type);
```

**Review by zero9178 — APPROVED** [review]
LGTM!

**Comment by joker-eph** [comment]
Please don't use **subscribers** teams to request reviews

**Comment by zero9178** [comment]
> Please don't use **subscribers** teams to request reviews

Could you elaborate what you mean? These are added automatically as they are code-owners of changed files

**Pull request merged**